cpp/include/cudf/ast/detail/expression_evaluator.cuh

```diff
@@ -277,48 +282,46 @@ struct expression_evaluator {
    * @return Element The type- and null-resolved data.
    */
   template <typename Element, CUDF_ENABLE_IF(column_device_view::has_element_accessor<Element>())>
-  __device__ possibly_null_value_t<Element, has_nulls> resolve_input(
-    detail::device_data_reference device_data_reference,
+  CUDA_DEVICE_CALLABLE possibly_null_value_t<Element, has_nulls> resolve_input(
+    detail::device_data_reference const& input_reference,
```

Comment:
Out of curiosity, did you try seeing if passing by `const` value had the same impact?

Reply:
I did test that, and it seemed to work in some cases but not others. In particular the nullable template had different performance characteristics from the non-nullable one, w.r.t to both this change and a number of others. The obvious culprit is the fact that the nullable template simply has more local state, so I need to reduce it even further to see measurable improvements in occupancy, but I also observed more subtle differences that I'd probably attribute to the compiler just not doing as good a job at some point. I have some ideas for redesigning the data references that might hopefully make these issues moot, but I'll need to play with those a bit and I figured that this perf boost was a nice short-term win while I do that.